dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/search/index/SearchPermissionsFinder.java

```diff
@@ -0,0 +1,207 @@
+import javax.inject.Inject;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Logger;
+
+/**
+ * Determine whether items should be searchable.
+ */
+@Stateless
+public class SearchPermissionsFinder {
+
+    private static final Logger logger = Logger.getLogger(SearchPermissionsFinder.class.getCanonicalName());
+
+    private RoleAssigneeServiceBean roleAssigneeService;
+    private DataverseRoleServiceBean rolesSvc;
+    
+    // -------------------- CONSTRUCTORS --------------------
+    
+    @Deprecated
+    public SearchPermissionsFinder() {
+        // JEE requirement
+    }
+    
+    @Inject
+    public SearchPermissionsFinder(RoleAssigneeServiceBean roleAssigneeService, DataverseRoleServiceBean rolesSvc) {
+        this.roleAssigneeService = roleAssigneeService;
+        this.rolesSvc = rolesSvc;
+    }
+
+    // -------------------- LOGIC --------------------
+    
+    /**
+     * @return A list of strings on which Solr will JOIN to enforce permissions
+     * @todo Should we make a PermStrings object? Probably.
+     */
+    public SearchPermissions findDataversePerms(Dataverse dataverse) {
+        List<String> permStrings = new ArrayList<>();
+        if (hasBeenPublished(dataverse)) {
+            return new SearchPermissions(permStrings, SearchPermissions.ALWAYS_PUBLIC);
+        }
+        
+        permStrings.addAll(findDvObjectPerms(dataverse));
+        return new SearchPermissions(permStrings, SearchPermissions.NEVER_PUBLIC);
+    }
+
+    public SearchPermissions findDatasetVersionPerms(DatasetVersion version) {
+        List<String> perms = new ArrayList<>();
+        if (version.isReleased()) {
+            return new SearchPermissions(perms, SearchPermissions.ALWAYS_PUBLIC);
+        }
+        
+        perms.addAll(findDvObjectPerms(version.getDataset()));
+        return new SearchPermissions(perms, SearchPermissions.NEVER_PUBLIC);
+    }
+    
+    /**
+     * Returns {@link SearchPermissions} for fileMetadata inside
+     * the given dataset version.
+     * <p>
+     * Note that {@link SearchPermissions} are specific to parent
+     * dataset version only.
+     * <p>
+     * This fact can be used to retrieve {@link SearchPermissions} for
+     * every file metadata inside single dataset version only once.
+     */
+    public SearchPermissions findFileMetadataPermsFromDatasetVersion(DatasetVersion version) {
+        Dataset dataset = version.getDataset();
+        List<String> perms = new ArrayList<>();
+        
+        Instant publicFrom = SearchPermissions.NEVER_PUBLIC;
+        
+        if (version.isReleased() && !dataset.getEmbargoDate().isDefined()) {
+            publicFrom = SearchPermissions.ALWAYS_PUBLIC;
+        } else if (version.isReleased()) {
+            publicFrom = dataset.getEmbargoDate()
+                    .map(d -> d.toInstant())
```

Comment:
I would treat lambda variables like normal variables, so I would give them meaningful names. Also this can be made simpler `.map(Date::toInstant)`.